Separating objects (#413)
the1lab/1lab: src/Cat/Diagram/Separator/Regular.lagda.md

  {o ℓ}
  (C : Precategory o ℓ)
  (coprods : (I : Set ℓ) → has-coproducts-indexed-by C ∣ I ∣)
  where
```

<!--
```agda
open Cat.Morphism.StrongEpi C
open Cat.Diagram.Separator.Strong C coprods
open Cat.Diagram.Separator C
open Cat.Reasoning C
open Copowers coprods

private variable
  s : Ob
```
-->

# Regular separators

:::{.definition #strong-separator}
Let $\cC$ be a category with [[set-indexed coproducts|indexed-coproduct]].
An object $S$ is a **regular separator** if the canonical map $\coprod_{\cC(S,X)} S \to X$
is a [[regular epi]].
:::

```agda
is-regular-separator : Ob → Type (o ⊔ ℓ)
is-regular-separator s = ∀ {x} → is-regular-epi C (⊗!.match (Hom s x) s λ e → e)
```

:::{.definition #strong-separating-family}
A family of objects $S_i$ is a **regular separating family** if the
canonical map $\coprod_{\Sigma(i : I), \cC(S_i, X)} S_i \to X$
is a [[regular epi]].
:::

```agda
is-regular-separating-family
  : ∀ (Idx : Set ℓ)
  → (sᵢ : ∣ Idx ∣ → Ob)
  → Type (o ⊔ ℓ)
is-regular-separating-family Idx sᵢ =
  ∀ {x} → is-regular-epi C (∐!.match (Σ[ i ∈ ∣ Idx ∣ ] (Hom (sᵢ i) x)) (sᵢ ⊙ fst) snd)
```

To motivate this definition, note that [[dense separators]] are extremely

> rare, as it imposes a strong discreteness condition on $\cC$. Instead, it

Reviewer: ```suggestion
rare, as they impose a strong discreteness condition on $\cC$. Instead, it
```

is slightly more reasonable to assume that every object of $\cC$ arises

> via some *quotient* of a bunch of points. It this intuition is what regular

Reviewer: ```suggestion
via some *quotient* of a bunch of points. This intuition is what regular
```

separators codify.

As the name suggests, regular separators and regular separating families

> are separators and separating families, resp. This follows directly

Reviewer: ```suggestion
are separators and separating families, respectively. This follows directly
```
